CPT: Add trash button to post type list item (v2)

aduth opened the pull request:
~~Cherry-picks from #5127 (00d4f315202c2e02f346cf8abc300f98ac430efe, 95d267d851dfeb5912dd2e49ef5e2edf5821d5f8)~~
Cherry-picks from #5007 (b0b0154c3ec2ab5f5f9e2f547622a98257880506)
Cherry-picks from #4557 (32a72102734e159d01f926df52930deca81cfd19, fdfdff29c67b3da6a533e25b9c408d714e93c774)

This pull request seeks to try an experimental alternative approach for dealing with post query state (and queryable state data generally) with better accommodation for updates to state over time. The included `QueryManager` class is intended to be extended by any use-case where we may need to track queried items and match, merge, and sort as revisions to that data are made in the application. The specific use-case in this pull request is in trashing a single post item. Refer to the "Open Questions" section of #4557 for more background on the complexities of this seemingly straight-forward interaction.

It's anticipated that if successful, this could be applied to terms state, which has similar requirements for filtering terms by search.

__Testing instructions:__

Ensure Mocha tests pass:

```
npm run test-client
```

Verify that you can trash posts whilst on the custom post type listing page. The post should instantly disappear from the post list when trashed.

For simplicity's sake, you can access the custom post types listing screen for the built-in posts type: http://calypso.localhost:3000/types/post

More interestingly, you can try variations where multiple queries are tracked simultaneously, noting that changing a post (e.g. trashing) will cause the item to appear in the newly applicable query. For example, navigating from "Trashed" filter to "Published", then trashing a post, if you were to navigate back to "Trashed", the new item should be reflected in the list with correct sorting, even before the REST API request has finished.

__Caveats:__

- There is not yet a notice to confirm that the post had been trashed successfully. 
- Post counts on the filter tabs do not yet update in accordance with changing status of post.

__Unknowns:__

- The continued existence of serialized query keys in the posts state reducer is somewhat undesirable (query request status, query last page). If these are necessary in other queryable state, we may want to seek a common approach.
- `queriesLastPage` post state would be affected by trashing posts, which is not reflected here

youknowriad commented:
@aduth great job here tho It seems to add some complexity in the reducers. AFAIK each time there's a new action responsible of updating something in posts, we should not forget to update the queries reducer to reflect those updates in the `QueryManager`.

What if use the same filters of the QueryManager here, but instead of updating the queries reducer on each action, we get rid of it completely ( apart from the loading state handling ) and update the `getSitePostsForQuery` selector (and all query based selectors) to use the `state.posts.items` as a starting point and apply the PostQueryManager rules as subselectors based on the query object, something like :

```javascript
function getSitePostsForQuery ( state, siteId, query ) {
  return getSitePosts( state, siteId )
    .filter(post => matchPostStats( post, query.status ))
    .filter(post => matchPostAfter( post, query.after ))
    .filter(post => matchPostBefore( post, query.before ))
// ....
   .splice(...) // pagination
   .sort(comparePostBy(query.order_by))
}
```
This seems more like the `redux` functional way (tho It's not really an argument).

The cons here could be performance, some profiling is needed but may be using `reselect` to combine the selectors could make the performance loss negligible.

What do you think about this ?

aduth commented:
@youknowriad : I'm very open to an idea like what you suggest. I think it would come down whether we're comfortable leaning entirely on the client to perform filtering of posts vs. as implemented currently where post query manipulation is a temporary state to provide the immediacy of a change's effects, but is replaced entirely by the REST API response on the next fetch. In other words, should we trust the client to perform this filtering? Or should the REST API response be the ultimate source of truth?

To your point on performance, we have our own implementation of selector memoization, [`lib/create-selector`](https://github.com/Automattic/wp-calypso/tree/master/client/lib/create-selector), so this should be a non-concern.

youknowriad commented:
@aduth yeah you're right about the choice of whether we rely on the backend or the front to perform the filtering. But some filtering is also done in front by the QueryManager.

Would you like me to try this approach in a separate PR ?

aduth commented:
>Would you like me to try this approach in a separate PR ?

I think it's worth experimenting, sure, but some worries and cautions worth thinking about:

- Is there any filtering we _can't_ do client-side that we'll likely want to use at some point in Calypso? Some list query arguments like `meta_key` weren't obvious to implement in `PostQueryManager` because there is server-side logic with respect to visibility of meta keys.
 - [Endpoint documentation](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/posts/)
 - [Endpoint implementation](https://github.com/Automattic/jetpack/blob/master/json-endpoints/class.wpcom-json-api-list-posts-endpoint.php)
- Is it possible to have some hybrid solution to let us do client-side filtering in selectors, but when a server-side set is received, to trust it as most accurate?
- Are issues around last page or key serialization as noted in the "Unknowns" of the original comment here better solved by a selector implementation?

timmyc commented on client/lib/query-manager/post/index.js:
```diff
@@ -0,0 +1,155 @@
+	 * Returns a new post after consideration of incoming revision of that
+	 * post. The post can be undefined in the case that the revision is
+	 * new. Optionally patch the post to merge, not replace. Returning
+	 * undefined indicates that post should be removed from known set. Posts
```
Why are CPT posts treated differently here?

aduth replied:
It's a bit of unfortunate legacy. Not sure the specific reasoning, but this is the behavior of core's `wp_delete_post`:

>wp_delete_post() automatically reverts to `wp_trash_post()` if $force_delete is false, the post_type of `$postid` is page or post, $postid is not already in the trash **and** if that trash feature enabled (which it it is by default).

https://codex.wordpress.org/Function_Reference/wp_delete_post

timmyc commented:
@aduth can you explain a bit around the need for `syncOptOut`?  My assumption is without this, the counts for a given post status get out of whack and/or a post appears in the incorrect status listing but just wanted to confirm.

timmyc commented:
Also post counts alongside the labels are obviously still wonky correct?

aduth commented:
>@aduth can you explain a bit around the need for `syncOptOut`?

It's a bit of cherry-picking from #5127, the standalone PR for the same changes (some more details around the need for it at that PR). Specifically, the sync handler is a bit buggy with how it works with promises, and due to the fact that we don't want to use the sync handler caching here* (vs. on the posts screen, where we'd like it to be used), I introduced the opt-out as a way to individually disable the default behavior.

_*_ Because we leverage the Redux state persistence for caching, so we have no need for two caching layers.

aduth commented:
>Also post counts alongside the labels are obviously still wonky correct?

That's correct. I thought I had noted that in the original comment, but it doesn't appear to be there. I'll update accordingly for clarity.

timmyc commented:
Thank you for the clarification.  This is testing out well for me ( with the known caveats around post counts ).  I did find a slight unrelated glitch that I fixed in #5294.  Following along with the discussion above with you and @youknowriad - I agree at face value this does seem to add some complexity in at the reducer level, but am in the camp of thinking the api should be the source of truth vs the client side.

Additionally we can re-utilize this in terms as well.

/cc @rralian since you have some history with the sync handler, and your feedback on disabling of the synch handler might be good to have here too.

I'm feeling good about this PR, but think it might be nice to have one other set of eyes on it prior to 🚢

rralian commented:
> /cc @rralian since you have some history with the sync handler, and your feedback on disabling of the synch handler might be good to have here too.

It was my suggestion, actually. I had learned through the sync handler project that the complexity of conflict resolution for infinite-lists is logarithmic in relation to the number of places where we store the data (an emotional calculation, not a literal one). The sync-handler was a hedge bet that we made so we could test out some offline functionality ahead of moving everything to redux. But to the extent that we can build new pieces in redux natively and make use of the redux persistence, they should avoid the sync-handler system altogether, IMHO.

timmyc commented:
After some additional discussion in slack with @aduth I think it would be good to move forward with this concept for now so we can get moving on additional CPT parts.  Obviously there are some caveats, but this is a working solution that will help with multiple areas of the project - and it does not affect any production areas yet.

gwwar commented on client/lib/query-manager/index.js:
```diff
@@ -0,0 +1,261 @@
+	 * Constructs a new instance of QueryManager
+	 *
+	 * @param {Object} data    Initial data
+	 * @param {Object} options Manager options
```
It'd be helpful here to list out the option properties and what they do.

aduth replied:
>It'd be helpful here to list out the option properties and what they do.
 
Added options property documentation to constructor and `receive` functions in rebased 25a39e2833b1e41f08dad263ec2620e032e4a308 ([line](https://github.com/Automattic/wp-calypso/pull/5135/commits/25a39e2833b1e41f08dad263ec2620e032e4a308#diff-90a76f3c723777b1a0c90ea618099ae8R38)).

gwwar commented on client/lib/query-manager/post/index.js:
```diff
@@ -0,0 +1,155 @@
+
+				case 'tag':
+				case 'category': {
+					const search = new RegExp( `^${ value }$`, 'i' );
```
Are we supporting regex search, or should this be escaped?

aduth replied:
>Are we supporting regex search, or should this be escaped?
 
No, we shouldn't support regex, only case-insensitive. Changes `search`, `tag`, and `category` to match on lowercase comparison in rebased 7c21ef133bb74fb2f8b684dbb75b6a3ec1b5442e ([line](https://github.com/Automattic/wp-calypso/pull/5135/commits/7c21ef133bb74fb2f8b684dbb75b6a3ec1b5442e#diff-4fe99528e32f35248cd67da8063b2ca5R32))

gwwar commented on client/state/posts/reducer.js:
```diff
@@ -47,12 +56,36 @@ export function items( state = {}, action ) {
+			// Posts and pages are first sent to trash before being permanently
+			// deleted. Therefore, only omit if already trashed or custom type.
+			const post = state[ globalId ];
+			if ( 'trash' === post.status || ! includes( [ 'post', 'page' ], post.type ) ) {
```
This feels odd that we don't have a separate trash action. I suppose `wpcom.site( siteId ).post( postId ).delete()` also mirrors this behavior?

aduth replied:
>This feels odd that we don't have a separate trash action. 

Indeed, this is unfortunately an artifact of the behavior of the REST API ([documentation](https://developer.wordpress.com/docs/api/1.1/post/sites/%24site/posts/%24post_ID/delete/))

gwwar commented:
👍 I took the branch out for a spin and it seems like a reasonable part of a solution to a very complex problem. 

> There is not yet a notice to confirm that the post had been trashed successfully.
> Post counts on the filter tabs do not yet update in accordance with changing status of post.

As we discussed, one approach to modifying posts/counts might very loosely look like the following:

1. Add a new reducer to store post deletion error state. We may need to generate a unique id per request so that we can keep track of multiple requests.
2. The component will then use a selector to keep track of if a post deletion has failed or succeeded.

```javascript
componentWillReceiveProps() {
    if ( this.props.deletePostFailed ) {
        showErrorNotice();
        incrementPostCount(); //reset counts
        clearDeletePostResponseState(); // clear redux error state
    }
    if ( this.props.deletePostSuccess ) {
        showSuccessNotice();
        clearDeletePostResponseState();
    }
}, 
clickDeletePost() {
    incrementPostCount(); //update counts
    deletePost(); // delete or trash a post
}
```

aduth commented:
It turns out we can't simply ignore pagination at this point in the implementation, as illustrated by this screenshot:

![queries state](https://cloud.githubusercontent.com/assets/1779930/15186902/dd85840c-176c-11e6-9cb3-50371bcd96d1.png)

Because the `match` function isn't capable of handling `page` query parameters, the second page of received results are incorrectly considered as matches in the first page results. In the screenshot above, the `{"type":"any"}` value should only have 20 items, since it's implicitly a query for the first page of results. The second page correctly contains 19 items, but only because matching logic is skipped for query results (received items simply assigned to query key). 

The consequence of this was that the `<PostSelector />` component, after having received the second page of results, was trying to render 58 items (39 from what it considered the first page, 19 from what it considered the second) with obvious duplication, leading to duplicate key warnings.

Two routes I'd considered in correcting this behavior:

- Update `match` to be capable of determining whether a post should be included in a query specifying `page` parameter. This seems difficult to implement (and perhaps non-performant) in light of the fact that page matching can only be determined after other query matching has concluded (i.e. non-matching items removed from set), followed by a check whether the position of the post in the entire set of sorted posts falls within the expected page range.
- Update the `PostQueryManager` implementation to ignore the `page` parameter (and any other parameter it's not capable of handling). I've implemented a working example of this in 1103530. On the upside, it simplifies post selectors by removing `IgnoringPage` suffixes. On the downside, we can't simply replace query results with those received from the server (since we're receiving paged results), so we may not always be confident it's reflective of the server-known set. Additionally, while we have no need for it currently, it wouldn't be possible to receive only the posts for a particular page (e.g. if we wanted to recreate truly paginated screens instead of infinite scroll).

I might plan to spend some more time trying to find a better compromise or alternative solution, but in the meantime 1103530 should fix the issue while satisfying all of our current requirements.

timmyc commented:
Good catch @aduth.

> if we wanted to recreate truly paginated screens instead of infinite scroll

We do have a classic paginated UI stats currently for listing followers

aduth commented:
I'm going to flip this back to In Progress pending further rethinking. Another case not well-covered here is how to handle fresh incoming data substituting existing data, as it relates to (a) accounting for items that may have been deleted or otherwise removed from the current query set, and (b) ensuring that infinitely scrollable lists refresh all of its pages, not just the first and whatever the "next" page is..

aduth commented:
This PR has haunted me for a while now, but I think I've addressed most of the challenges we've faced. I'd experimented with a number of alternative solutions, but decided to stick with the `QueryManager` class approach. There's quite a bit more (necessary) added complexity this time around, but in introducing a new `PaginatedQueryManager` class from which `PostQueryManager` extends, it's now capable of handling the redistribution of pages after a change is made. It also understands "found" items for a query, and internally adjusts the count as changes cause items to be added or removed (so that infinite scroll last page is kept accurate).

The [test scenarios](https://github.com/Automattic/wp-calypso/pull/5135/files#diff-b3303ef5ef84f8a863d4da92403564b8R135) are a good demonstration of how this works.

More notes:

- I've temporarily removed persistence of posts queries state. One of the remaining challenges is re-fetching paginated data where items may be removed. While `PaginatedQueryManager` can handle this here already, the current infinite list components (both `<InfiniteList />` and `<PostSelector />`) aren't capable of knowing what pages they're already aware of (e.g. if we were to restore 4 pages of post data from state). My plan for this is to update these components in a subsequent PR to call a function prop when the visible range of items changes and render query components from the parent component for pages which should be requested.
- I may follow-up with another PR to consolidate `items` state, which is effectively duplicated between the query manager and `state.posts.items`. Eliminating `state.posts.items` and leaning more on `PostQueryManager` would simplify posts state, at the cost of losing the ability to look up posts efficiently by global ID (not convinced we're making use of this, though).
- I've changed the serialization of queries, which was previously a relatively naive `JSON.stringify` call. The issue with this is that `{ a: 1, b: 2 }` would be serialized differently than `{ b: 2, a: 1 }`, even though they're effectively the same object (insofar as it would affect a query). Now it's converted to a sorted pairs array using [`_.toPairs`](https://lodash.com/docs#toPairs) before being stringified. (I also found [`json-stable-stringify`](https://github.com/substack/json-stable-stringify) as an alternative)

apeatling commented:
@gwwar @timmyc Could you have a final look at this one so we can move this forward?

gwwar commented:
👍 I took the branch out for another spin and didn't see any unreported issues. I'm going to do a deeper code 🔍 on Monday though since this is a very large change set. 

>  Eliminating state.posts.items and leaning more on PostQueryManager would simplify posts state, at the cost of losing the ability to look up posts efficiently by global ID

I wonder how would we reference posts outside of the posts tree. Would we still use a global id and memoize the lookup?

> The issue with this is that { a: 1, b: 2 } would be serialized differently than { b: 2, a: 1 }

⭐ Thanks for pointing this one out

gwwar commented on client/lib/query-manager/key.js:
```diff
@@ -0,0 +1,32 @@
+	 * @return {String}       Serialized query
+	 */
+	static stringify( query ) {
+		return JSON.stringify( sortBy( toPairs( query ), ( pair ) => pair[ 0 ] ) );
```
Would it be possible to use the following, or would it still be sometimes incorrect after a sort?

```javascript
return JSON.stringify( fromPairs( sortBy( toPairs( query ), ( pair ) => pair[ 0 ] ) ) );
```

aduth replied:
>Would it be possible to use the following, or would it still be sometimes incorrect after a sort?

While in practice it would probably work, I'm quite sensitive to the performance here (since this is run quite frequently) and I'm not too concerned that the serialized key look like JSON.

gwwar commented:
👍 The code here looks good, though it's very likely that I missed things in this large change set. 

@timmyc did you get a chance to have a 👀

timmyc commented:
I have given this one quite a bit of 👀  and the tests have been very helpful to read/grok what is going on.  I think we should get this merged in again so we can work with it, bend it a bit, and look for issues while getting the rest of CPT dialed in.

Again since this only affects non-production code, and with the test coverage to boot, I think we should merge away.

aduth commented:
>Again since this only affects non-production code, and with the test coverage to boot, I think we should merge away.

While I wish this were the case, it does affect production in that the posts state is used by the posts selector in the editor (parent page, link to existing content).

That said, assuming code looks good and works well, shouldn't hold up a merge. Thanks for taking a look!

aduth commented:
Many thanks @gwwar and @timmyc for taking on the task of reviewing this behemoth. I've spent the morning doing a final pass over the changes and testing all affected sections, and I think this one is finally ready to merge.
